### PR TITLE
feat: expose flow readiness check in gateway

### PR DESCRIPTION
**Pull request by JoanFM**
This PR aims to expose a `health` endpoint to the gateway that works as a dry_run to expose to the client a way to know if the Flow is ready to receive requests. 

It is important to differentiate this `health` endpoint with the `healthiness check` of an individual runtime.

Maybe the name `health` or `health_check` should be changed in this context to some `dry_run`?

### CODE REVIEW

File: jina/serve/executors/__init__.py [hunk 1]
```diff
@@ -110,6 +112,16 @@ def __init__(
...
+        if __dry_run_endpoint__ not in self.requests:
+            self.requests[__dry_run_endpoint__] = self._dry_run_func
+        else:
+            self.logger.warning(
+                f' Endpoint {__dry_run_endpoint__} is defined by the Executor. Be aware that this endpoint is usually reserved to enable health checks from the Client through the gateway.'
+                f' So it is recommended not to expose this endpoint. '
+            )
+
+    def _dry_run_func(self, *args, **kwargs):
+        pass
```
Comment (samsja) [inline jina/serve/executors/__init__.py:124]: Should not we log something here (in debug level )  ?

File: jina/serve/runtimes/gateway/grpc/__init__.py [hunk 2]
```diff
@@ -127,3 +130,34 @@ async def async_run_forever(self):
...
+        da = DocumentArray()
+
+        try:
+            req_iterator = request_generator(
+                exec_endpoint=__dry_run_endpoint__,
+                data=da,
+                data_type=DataInputType.DOCUMENT,
+            )
+            async for _ in self.streamer.stream(request_iterator=req_iterator):
+                pass
```
Comment (samsja) [inline jina/serve/runtimes/gateway/grpc/__init__.py:156]: Same as above maybe log something here. The goal would be that if someone want to understand where something is failing he can check from the log that the health check as been send from the gateway but not receive in the Executor

File: jina/serve/runtimes/gateway/http/app.py [hunk 3]
```diff
@@ -102,17 +102,54 @@ async def _shutdown():
...
             path='/',
-            summary='Get the health of Jina service',
+            summary='Get the health of Jina gateway service',
             response_model=JinaHealthModel,
         )
-        async def _health():
+        async def _gateway_health():
             """
-            Get the health of this Jina service.
+            Get the health of this gateway service.
```
Comment (samsja) [inline jina/serve/runtimes/gateway/http/app.py:110]:
```suggestion
            Get the health of this Gateway service.
```

**Review by samsja — COMMENTED** [review]
This feature is really great ! I really like the choice of doing the dry run on a additional endpoint

File: jina/serve/executors/__init__.py (hunk 1 above)
Comment (JoanFM) [reply jina/serve/executors/__init__.py:124]: in theory u will see `DataRequest on `__dry_run_endpoint__` received

Comment (samsja) [reply jina/serve/executors/__init__.py:124]: yes you are right

**Comment by JohannesMessner** [comment]
> Maybe the name `health` or `health_check` should be changed in this context to some `dry_run`?

IMO having the endpoint for the dry run be called `dry_run` makes the most sense

File: docs/fundamentals/health-check.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+{
+  "status": "SERVING"
+}
+```
+
+(health-check-gateway)=
+# Health check of the Gateway
+
+The same way individual Executors expose endpoints for orchestrators, clients or other services to check their availability, Gateway as a microservice also exposes this in different ways and depending
+on the protocol used.
```
Comment (samsja) [inline docs/fundamentals/health-check.md:35]:
```suggestion
The same way individual Executors expose endpoints for orchestrators, clients or other services to check their availability, Gateway, as a microservice, also exposes this in different ways depending on the protocol used.
```

File: docs/fundamentals/health-check.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+
+## Gateway health check with grpc
+
+When using grpc as the protocol used to communicate with the Gateway, then Gateway uses the exact same mechanism as Executors to expose their individual health status. It exposes [grpc health check](https://github.com/grpc/grpc/blob/master/doc/health-checking.md) to the outside world.
```
Comment (samsja) [inline docs/fundamentals/health-check.md:40]:
```suggestion
When using grpc as the protocol to communicate with the Gateway, then Gateway uses the exact same mechanism as Executors to expose their individual health status. It exposes [grpc health check](https://github.com/grpc/grpc/blob/master/doc/health-checking.md) to the outside world.
```

File: docs/fundamentals/health-check.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+```
+
+Then you will get a valid empty response indicating its ability to serve.
+```text
+{}%
+```
+
+# Readiness of a Flow exposed to the client
+
+A lot of times, from the client perspective, it is useful to know if a Flow as a complete set of microservices is ready to receive requests. This is why the gateway
```
Comment (samsja) [inline docs/fundamentals/health-check.md:80]:
```suggestion
A lot of times, from the client perspective, it is useful to know if a Flow, as a complete set of microservices, is ready to receive requests. This is why the gateway 
```

File: docs/fundamentals/health-check.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+DEBUG  executor0/rep-0@19074 start listening on 0.0.0.0:12346                                                                                                                                   [05/31/22 18:10:16]
+DEBUG  gateway/rep-0@19075 start server bound to 0.0.0.0:12345                                                                                                                                  [05/31/22 18:10:17]
+DEBUG  executor0/rep-0@19059 ready and listening                                                                                                                                                [05/31/22 18:10:17]
+DEBUG  gateway/rep-0@19059 ready and listening                                                                                                                                                  [05/31/22 18:10:17]
+╭────── 🎉 Flow is ready to serve! ──────╮
+│  🔗  Protocol                  GRPC    │
+│  🏠     Local         0.0.0.0:12345    │
+│  🔒   Private    192.168.1.13:12345    │
+│  🌍    Public  92.178.210.177:12345    │
```
Comment (samsja) [inline docs/fundamentals/health-check.md:147]:
```suggestion
```

Maybe you don't want you personnal IP to be on the documentation

**Review by samsja — COMMENTED** [review]
I like the structure of the docs and it is really didactic.
Probably @JohannesMessner will have some things to say

**Pull request merged**